### PR TITLE
custom_target: Add env kwarg

### PR DESCRIPTION
**Pull request by xclaesse**

### CODE REVIEW

**Comment by tristan957** [comment]
Will env valeus have @KEYWORD@ substitution like the arguments do?

File: docs/markdown/Reference-manual.md [hunk 1]
```diff
@@ -372,6 +372,10 @@ following.
 - `install_mode` *(since 0.47.0)*: the file mode and optionally the
   owner/uid and group/gid
 - `output`: list of output files
+- `env` *(since 0.57.0)*: environment variables to set, such as `['NAME1=value1',
+   'NAME2=value2']`, or an [`environment()`
+   object](#environment-object) which allows more sophisticated
+   environment juggling. A dictionary is also accepted.
```
Comment (dcbaker) [inline docs/markdown/Reference-manual.md:378]: since dicts are always supported, maybe reword this the opposite way?
```markdown
environment variables to set, such as `{'NAME1': 'value1', 'NAME2': 'value2'}`, or ... A list of strings in the form `NAME=value` is also accepted.
```

Comment (dcbaker) [reply docs/markdown/Reference-manual.md:378]: I really feel like in the long run we probably want to deprecate the `['NAME=value']` syntax, as it's pretty magical and dicts or `environment()` objects are much nicer to work with anyway.

**Review by dcbaker — APPROVED** [review]
LGTM. One suggestion on the wording of the documentation, take it or leave it.

**Comment by xclaesse** [comment]
> Will env valeus have @Keyword@ substitution like the command arguments do?

Not atm, could be added later I guess if needed? What's your use-case?

File: docs/markdown/Reference-manual.md (hunk 1 above)
Comment (xclaesse) [reply docs/markdown/Reference-manual.md]: Right, better have the dict syntax first since that's what we should recommend now, made that change now. We could indeed deprecate the string list format, but I preferred to keep it here for consistency with other APIs.

**Comment by tp-m** [comment]
Yay!

**Comment by tristan957** [comment]
> > Will env valeus have @Keyword@ substitution like the command arguments do?
> 
> Not atm, could be added later I guess if needed? What's your use-case?

Was just thinking to be consistent with arguments.

**Comment by tristan957** [comment]
@xclaesse could this be extended to run_target() as well?

Use case: I am building a Python extension, and want to add a run_target() that opens a Python shell with the PYTHONPATH pointing to the correct folder in the builddir.

**Comment by xclaesse** [comment]
That would make sense, and made me realize run_target() uses completely different code path than custom_target() to run the command. Big code duplication that should get fixed too.

**Comment by tristan957** [comment]
Cool. Code clean-up is a beautiful thing.

**Comment by xclaesse** [comment]
@tristan957 FYI it's in https://github.com/mesonbuild/meson/pull/8305

**Comment by tristan957** [comment]
Did this make it in for 0.57? Seems like 0.57 could be a fun release

**Comment by xclaesse** [comment]
For run_target() it's not yet merged, not sure we'll have time since the freeze is this weekend. Unfortunately since it's a much bigger refactor than expected, it could have to be delayed to 0.58. Depends if someone can do the review...

**Comment by tristan957** [comment]
That's cool. Nbd either way
